src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -61,97 +66,56 @@ class ResourceRequest {
+    if (it == resources_.end()) {
       return FixedPoint(0);
     } else {
-      return *ptr;
+      return it->second;
     }
   }
 
   /// Set a resource to the given value.
   /// NOTE: if the new value is 0, the resource will be removed.
   ResourceRequest &Set(ResourceID resource_id, FixedPoint value) {
     if (value == 0) {
-      if (IsPredefinedResource(resource_id)) {
-        predefined_resources_[resource_id.ToInt()] = 0;
-      } else {
-        custom_resources_.erase(resource_id.ToInt());
-      }
+      resources_.erase(resource_id);
     } else {
-      auto ptr = GetPointer(resource_id);
-      if (ptr == nullptr) {
-        custom_resources_[resource_id.ToInt()] = value;
-      } else {
-        *ptr = value;
-      }
+      resources_[resource_id] = value;
     }
     return *this;
   }
 
   /// Check whether a particular resource exist.
-  bool Has(ResourceID resource_id) const {
-    auto ptr = GetPointer(resource_id);
-    return ptr != nullptr && *ptr != 0;
-  }
+  bool Has(ResourceID resource_id) const { return resources_.contains(resource_id); }
 
   /// Clear the whole set.
-  void Clear() {
-    for (size_t i = 0; i < predefined_resources_.size(); i++) {
-      predefined_resources_[i] = 0;
-    }
-    custom_resources_.clear();
-  }
+  void Clear() { resources_.clear(); }
 
   /// Remove the negative values in this set.
   void RemoveNegative() {
-    for (size_t i = 0; i < predefined_resources_.size(); i++) {
-      if (predefined_resources_[i] < 0) {
-        predefined_resources_[i] = 0;
-      }
-    }
-    for (auto it = custom_resources_.begin(); it != custom_resources_.end();) {
+    for (auto it = resources_.begin(); it != resources_.end();) {
       if (it->second < 0) {
-        custom_resources_.erase(it++);
+        resources_.erase(it++);
       } else {
         it++;
       }
     }
   }
 
   /// Return the number of resources in this set.
-  size_t Size() const {
-    size_t size = custom_resources_.size();
-    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
-      if (predefined_resources_[i] != 0) {
-        size++;
-      }
-    }
-    return size;
-  }
+  size_t Size() const { return resources_.size(); }
 
   /// Return true if this set is empty.
-  bool IsEmpty() const { return Size() == 0; }
-
-  /// Return a set that contains all resource ids in this set.
-  absl::flat_hash_set<ResourceID> ResourceIds() const {
-    absl::flat_hash_set<ResourceID> res;
-    for (size_t i = 0; i < predefined_resources_.size(); i++) {
-      if (predefined_resources_[i] != 0) {
-        res.insert(ResourceID(i));
-      }
-    }
-    for (auto &entry : custom_resources_) {
-      res.insert(ResourceID(entry.first));
-    }
-    return res;
-  }
+  bool IsEmpty() const { return resources_.empty(); }
+
+  /// Return a boost::range object that can be used as an iterator of the resource IDs.
+  ResourceIdIterator ResourceIds() const { return boost::adaptors::keys(resources_); }
```

Comment:
what happens to the iterator if the resources_ is changed?

Reply:
Good question. I could find the documentation about this. 
But I did 2 tests. They work well.
```
Test 1:
  r1 = ResourceRequest({{cpu_id, -1}, {custom_id1, 2}, {custom_id2, -2}});
  for (auto &resource_id : r1.ResourceIds()) {
    r1.Set(resource_id, 0);
    RAY_LOG(INFO) << resource_id;
  }
  
Prints:
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: CPU
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: custom2
[2022-03-31 08:32:47,385 I 185337 185337] resource_request_test.cc:158: custom1

Test 2:
  r1 = ResourceRequest({{cpu_id, -1}, {custom_id1, 2}, {custom_id2, -2}});
  for (auto &resource_id : r1.ResourceIds()) {
    r1.Clear();
    RAY_LOG(INFO) << resource_id;
  }

Prints:
[2022-03-31 08:23:29,366 I 180899 180899] resource_request_test.cc:158: custom2


```